### ISSUE TITLE
Reinterpret immutables

### ISSUE THREAD

**Issue by SimonDanisch** [JuliaLang/julia#10140]
I was wondering, why this works:
```Julia
reinterpret(Float32, Int32(0))
@assert isbits(ImmutableType1) && isbits(ImmutableType2) && (sizeof(ImmutableType1) == sizeof(ImmutableType2)) 
reinterpret(ImmutableType1, Array(ImmutableType2, 42))
```
and this does not:
```Julia
reinterpret(ImmutableType1, ImmutableType2(...))
```
This is convenient for converting different FixedSizeArrays with the same layout without copying anything.
It's probably related to #7697, which didn't get fully implemented?

@vtjnash answered on Julia users:
> That last one requires a different sequence of llvm instructions than the other examples (it's the difference between isbits and bitstype). Not hard, but likely unimplemented. 

Is anyone willing to implement this, or give some pointers on how to implement it?

**Comment by JeffBezanson** [comment]
Look for `generic_box` in intrinsics.cpp. It uses `jl_is_bitstype`, which checks for types declared with `bitstype`. Changing this to `jl_isbits` would make it work for more types. The same change would probably be needed in a couple other places in that file, such as `try_to_determine_bitstype_nbits`.

**Comment by vtjnash** [comment]
There is an actual distinction between bitstype and isbits which becomes highly significant when dealing with llvm and C interop.

But Keno and I have written the necessary code for this dozens of times (for ccall and getfield), so I should perhaps just throw it in with all the other ccall updates.

**Comment by JeffBezanson** [comment]
Reinterpreting between two isbits types is trivial compared to handling ccall; you can clearly have the former independent of the latter.

**Comment by vtjnash** [comment]
yes, that is clearly true. i meant the reverse: implementing ccall has required implementing this operation several times

**Comment by vtjnash** [comment]
it was decided that immutable objects are distinct from bitstypes. the former are not expected to be constructible via direct memory casts as this raises issues of padding, alignment, and ignoring the constructor.

**Comment by timholy** [comment]
Thanks to whoever was involved for taking the time to discuss and come to a conclusion. These are never simple decisions.

Since "padding" and "alignment" are scary words, am I right in hoping that `reinterpret` is still OK for an array with immutable elements? E.g.,
```jl
julia> immutable TwoInts
           a::UInt16
           b::UInt16
       end

julia> a = TwoInts(1,2)
TwoInts(0x0001,0x0002)

julia> reinterpret(Float32, a)
ERROR: reinterpret: expected bitstype value for second argument
 in reinterpret(::Type{Float32}, ::TwoInts) at ./essentials.jl:84
 in eval(::Module, ::Any) at ./boot.jl:243

julia> reinterpret(Float32, [a])
1-element Array{Float32,1}:
 1.83672e-40
```
If that's somehow OK (even if the scalar case is not), then would that make
```jl
julia> reinterpret(Float32, Ref(a))
ERROR: reinterpret: expected bitstype value for second argument
 in reinterpret(::Type{Float32}, ::Base.RefValue{TwoInts}) at ./essentials.jl:84
 in eval(::Module, ::Any) at ./boot.jl:243
```
also safe? It still bypasses the constructor, so I can understand why you might not want to make this easy. But for those who need/want to use such constructs, I guess the question is, "how dangerous is this?"

Obviously I don't care about `TwoInts` specifically, I just mean whether it's safe in general.

An example of a use case is saving/loading an array of immutables to disk via HDF5/JLD; it's basically going to get packed as an `Array{UInt8,1}` and then `reinterpreted` as the correct type. As far as I can see, there really isn't much of a way around this, since it involves interaction between julia and C, and C doesn't have adequate respect :smile: for julia types.

**Comment by timholy** [comment]
In other words, if despite my impression arrays are also not entirely safe, at least I'd rather know about it, even if there isn't currently a good alternative.

**Comment by vtjnash** [comment]
`Array{UInt8}` is a bit of a special case that doesn't really fall out of the definition of `reinterpret(Type, Val)`, but is more closely related to the equivalent C `void*` operation (with bounds-checking). I think it might make additional sense to limit conversions to need to go through byte vectors (so you couldn't treat `Array{Complex128}` as `Array{Float64}`, for example). If we're expecting to support general aliasing memory casts of arrays, we're going to need to be very careful how we mark the TBAA for accesses to that data (once we start doing that marking). One other issue I forgot to mention above is that the direct cast example above has endian-dependent behavior while the equivalent explicit load and shift does not (ref discussion #13701).

fwiw, in #14766 I added a new primitive I/O functions unsafe_read and unsafe_write for copying from `nbytes` from a `Ptr{Void}` into an object. the primary goal was to make it possible to perform I/O on the bytes of a random object without requiring `reinterpret`. this helps make it explicit that these operations are bypassing the normal constructors. adding a custom reader/writer can be commonly expressed using a simple opt-in function definition: `read(io, MyType) = Base.unsafe_read(io, Ref{MyType})[]`.

**Comment by vtjnash** [comment]
tl;dr; it's safe for the common workflows, but there may be opportunities here to support the desired operations with more careful primitives

**Comment by timholy** [comment]
Thanks for the pointer on `unsafe_read`. Does it handle endianness, or is that a separate problem?

**Comment by vtjnash** [comment]
endianness is separate. we could add it add something to base, but I think it may be hard to define the general case (is it a bswap over the fields or a bswap over the fields in the field?)

**Comment by timholy** [comment]
> I think it might make additional sense to limit conversions to need to go through byte vectors (so you couldn't treat Array{Complex128} as Array{Float64}, for example).

As you make decisions about future paths, here's another data point for you. A number of people who do image processing seem to like to go back and forth between `Array{RGB{Float32}, 2}` and `Array{Float32, 3}` with a first dimension of size 3. This may be a holdover from prior programming experience (I don't do this much myself), but I'm not sure. For example, since 3-vectors don't have an order, `maximum(A)` doesn't make sense for an `Array{RGB}`, but `maximum(A, (2,3))` gives you a natural way to calculate the maximum in each color channel when `A::Array{Float32,3}`.

**Comment by eschnett** [comment]
It is very natural to reinterpret an immutable type that has fields of a single type as `NTuple`, or as array. The underlying mathematical construct is a vector space -- there's no real Julia concept that represents a vector space, but `Array` certainly comes closest.

I wish there was a simple way to say that e.g. `NTuple` is a subtype of `AbstractArray`, or that an immutable type is equivalent to a tuple, etc.

This comes up all the time also when one discretizes PDEs, which uses algorithms that are often similar to image processing.

I think the long-term solution is to introduce a concept ("trait") to Julia that captures the idea of a vector space, and have this integrated well enough with the  type system that it is not necessary to reinterpret immutables or tuples this way. I would hope that using a trait gives you the same expressive power as "subtype of `AbstractArray`", but be as efficient as hand-written code.

In the mean time, being able to reinterpret types in such a way that something in Julia's run-time system ensures that the type layouts match would be great.

**Comment by vtjnash** [comment]
only allowing reinterpret if the ordering and types of the underlying bitstypes are the same sounds likely to be valid, and shouldn't require pessimizing tbaa too significantly. i don't plan on making any changes to this in the near future, but these are good examples to consider

**Comment by bjarthur** [comment]
it is still not clear to me what the underlying trouble with reinterpreting a scalar is.  if conceptually we've figured out how to make it works for arrays, what breaks with a scalar?

**Comment by robsmith11** [comment]
I'm wondering the same as @bjarthur.  Why can't `reinterpret` operate on scalar structs without wrapping them in arrays?

**Comment by vtjnash** [comment]
It could (read the discussion above). It wasn't done because there seem to be faster, simpler, and/or more efficient alternatives (or with fewer gotchas) that we hope to push users towards.

**Comment by vtjnash** [comment]
And in reply to bjarthur's question: because we have _not_ figured out arrays (https://github.com/JuliaLang/julia/issues?utf8=✓&q=is%3Aissue+is%3Aopen+reinterpretarray), and in fact have entirely replaced the implementation that was in place at the time of his query

**Comment by robsmith11** [comment]
> It could (read the discussion above). It wasn't done because there seem to be faster, simpler, and/or more efficient alternatives (or with fewer gotchas) that we hope to push users towards.

@vtjnash Thanks, I reread the discussion, but still am unclear on what the faster/simpler alternatives are.  I see `unsafe_read` mentioned, but that seems to operate on IO streams.  How would that be used as a solution to tim's `TwoInts` to `Float32` example?
